### PR TITLE
feat: more robust Linux cross-user cancelation under low memory conditions

### PR DESCRIPTION
**Pull request by jusiskin**
### What was the problem/requirement? (What/Why)

Under low memory situations on Linux, the worker agent can fail to cancel a running session action.

On Linux, processes can only send signals to other processes if at least one of the following are true:

*   the target process is owned by the same OS user as the signalling process OR
*   if the process sending the signal
    *   is `root`
    *   has the `CAP_KILL` capability (see[capabilities(7) man page](https://man7.org/linux/man-pages/man7/capabilities.7.html))

To overcome this limitation, `openjd-sessions` used `sudo` to create a `bash` process as the target user which then sends the OS signal. When the system has low memory, there may not be sufficient memory to create the subprocesses successfully.

### What was the solution? (How)

OpenJobDescription/openjd-sessions-for-python#196 is a work-in-progress enhancement to leverage `CAP_KILL` and send signals to cross-user processes without creating subprocesses.

This PR builds on this support for `CAP_KILL` so the worker agent directly signals cross-user processes.

1.  The code for `install-deadline-worker` was modified to configure the worker agent systemd unit with the `CAP_KILL` ambient capability. This has the effect of adding `CAP_KILL` to the worker agent process' permitted/effective/inheritable capability sets.
2. The worker agent now drops `CAP_KILL` from the inheritable capability set early in the program startup so that other threads and subprocesses will not inherit this privileged capability.
3. The `openjd-session` change takes care of the rest and handles direct signalling for cross-user processes.

### What is the impact of this change?

This change allows the worker agent to cancel session actions more robustly in low-memory situations.

### How was this change tested?

The cross-user direct signal cancelation tests live in `openjd-sessions`, but a security end-to-end test was added to ensure that session actions are not able to signal other processes.

### Was this change documented?

No

### Is this a breaking change?

No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline_worker_agent/installer/install.sh [hunk 1]
```diff
@@ -456,15 +456,28 @@ Environment=AWS_REGION=$region AWS_DEFAULT_REGION=$region
 EOF
     fi
     
+    ###############################################################
+    ###############    NOTE FOR CODE REVIEWERS    #################
+    ###############################################################
+    #    Review changes to AmbientCapabilities below carefully    #
+    ###############################################################
+
```
Comment (ddneilson) [inline src/deadline_worker_agent/installer/install.sh:464]: Nice touch. I like it!

File: src/deadline_worker_agent/linux/capabilities.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        raise OSError(errno, os.strerror(errno))
+    return result
+
+
+@cache
+def _get_libcap() -> ctypes.CDLL:
+    if not sys.platform.startswith("linux"):
+        raise OSError(f"libcap is only available on Linux, but found platform: {sys.platform}")
+
+    libcap = ctypes.CDLL(find_library("cap"), use_errno=True)
```
Comment (ddneilson) [inline src/deadline_worker_agent/linux/capabilities.py]: Same issue as the `openjd-sessions` PR. We're assuming that the lib is always present. We modified the other change to account for it not being present.

File: test/e2e/test_cap_kill.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    # GIVEN
+    # Send SSM command to write and run a bash script
+    # The script creates a detached sleep process and outputs that process' PID
+    # This sleep process will run as the ssm-user which is different from the job user
+    result = session_worker.send_command(
+        " ; ".join(
+            [
+                "echo '#!/bin/bash' > script.sh",
+                "echo 'set -euo pipefail' >> script.sh",
+                "echo 'nohup sleep 120 < /dev/null 2> /dev/null > /dev/null &' >> script.sh",
```
Comment (ddneilson) [inline test/e2e/test_cap_kill.py]: Two questions:
1. Is 120s sufficient? I'm thinking about potential test timing. e.g. if it, for some reason, takes longer than expected for the worker to pick up and run the submitted job.
2. Should we have an ssm command at the end of the test that explicitly kills the script, just to make sure that we're not leaving anything behind? There's *probably* no harm, right now, in the test leaving a script running for a short time, but if this becomes a pattern that is replicated then the replicant may clash with this test in random unexpected ways...

**Review by ddneilson — CHANGES_REQUESTED** [review]

**Comment by ddneilson** [comment]
Oh, also, the README should probably be updated to list the (optional) dependency on `libcap.so` on linux

File: src/deadline_worker_agent/linux/capabilities.py (hunk 2 above)
Comment (jusiskin) [reply src/deadline_worker_agent/linux/capabilities.py]: I've added error handling for when `libcap` cannot be found. On Amazon Linux, it appears that systemd depends on libcap:

```sh
$ dnf repoquery -q --installed --whatrequires libcap
chrony-0:4.3-1.amzn2023.0.4.x86_64
coreutils-0:8.32-30.amzn2023.0.3.x86_64
iproute-0:6.10.0-319.amzn2023.0.1.x86_64
iputils-0:20210202-2.amzn2023.0.4.x86_64
kernel-tools-0:6.1.115-126.197.amzn2023.x86_64
libcap-devel-0:2.48-2.amzn2023.0.3.x86_64
nfs-utils-1:2.5.4-2.rc3.amzn2023.0.3.x86_64
rpm-build-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-plugin-selinux-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-plugin-systemd-inhibit-0:4.16.1.3-29.amzn2023.0.6.x86_64
rpm-sign-libs-0:4.16.1.3-29.amzn2023.0.6.x86_64
systemd-0:252.23-2.amzn2023.x86_64
systemd-libs-0:252.23-2.amzn2023.x86_64
systemd-pam-0:252.23-2.amzn2023.x86_64
```

In the ["Worker host" section Deadline Cloud user guide](https://docs.aws.amazon.com/deadline-cloud/latest/userguide/worker-host.html), we recommend using Amazon Linux:

>  To complete the tasks in the following sections, you need the following:
>   *   A Linux-based Amazon Elastic Compute Cloud (Amazon EC2) instance. We recommend Amazon Linux 2023.

systemd is used to launch the worker agent on boot and restart on a crash. We also configure the worker agent to be given `CAP_KILL` in the systemd unit generated by `install-deadline-worker`.

On this basis, we decided to log a warning when the library is not found and move on. Those running the worker agent on Linux with `CAP_KILL` but without `libcap` available is not officially supported.

File: test/e2e/test_cap_kill.py (hunk 3 above)
Comment (jusiskin) [reply test/e2e/test_cap_kill.py]: I've bumped the sleep to 240s which should be sufficient. I've also moved the background sleep job to a pytest fixture which handles cleanup of the process

**Review by AWS-Samuel — APPROVED** [review]

**Review by ddneilson — APPROVED** [review]
Thanks for taking this on, Josh. Looks great!

**Pull request merged**